internal/cache/devices.go

```diff
@@ -25,24 +39,64 @@ type DeviceCache interface {
 	Update(device models.Device) errors.EdgeX
 	RemoveByName(name string) errors.EdgeX
 	UpdateAdminState(name string, state models.AdminState) errors.EdgeX
+	SetLastConnectedByName(name string)
+	GetLastConnectedByName(name string) int64
 }
 
 type deviceCache struct {
-	deviceMap map[string]*models.Device // key is Device name
-	mutex     sync.RWMutex
+	deviceMap     map[string]*models.Device // key is Device name
+	mutex         sync.RWMutex
+	dic           *di.Container
+	lastConnected map[string]gometrics.Gauge
 }
 
-func newDeviceCache(devices []models.Device) DeviceCache {
+func newDeviceCache(devices []models.Device, dic *di.Container) DeviceCache {
 	defaultSize := len(devices)
 	dMap := make(map[string]*models.Device, defaultSize)
 	for i, d := range devices {
 		dMap[d.Name] = &devices[i]
 	}
 
-	dc = &deviceCache{deviceMap: dMap}
+	dc = &deviceCache{deviceMap: dMap, dic: dic}
+	lastConnectedMetrics := make(map[string]gometrics.Gauge)
+	for _, d := range devices {
+		deviceMetric := gometrics.NewGauge()
+		registerMetric(d.Name, deviceMetric, dic)
+		lastConnectedMetrics[d.Name] = deviceMetric
+	}
+	dc.lastConnected = lastConnectedMetrics
+
 	return dc
 }
 
+func registerMetric(deviceName string, metric interface{}, dic *di.Container) {
+	metricsManager := bootstrapContainer.MetricsManagerFrom(dic.Get)
+	lc := bootstrapContainer.LoggingClientFrom(dic.Get)
+	configuration := container.ConfigurationFrom(dic.Get)
+	if configuration.Service.EnableNameFieldEscape {
+		deviceName = common.URLEncode(deviceName)
+	}
```

Comment:
Not sure we need to be doing this here. Currently there are no restrictions on the metric name other then non-empty string. IMO the escaping needs to be handled in the Reporter here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/bootstrap/metrics/reporter.go#L186

Reply:
Let's discuss this with more details on the device WG meeting tomorrow, thanks Lenny.

Reply:
As Lenny indicated in the meeting, the device name only is included in the metric message payload, removed the `EnableNameFieldEscape` check here as the full publish topic name is `edgex/telemetry/device-simple/LastConnected`.

@lenny-intel change's done and is ready for review, thanks!